src/coprocessor/statistics/analyze.rs

```diff
@@ -271,6 +291,230 @@ impl<S: Snapshot> RequestHandler for AnalyzeContext<S> {
+                collector.collect_column(column_vals);
+            }
+        }
+        Ok(AnalyzeSamplingResult::new(collector))
+    }
+}
+
+#[derive(Clone)]
+struct RowSampleCollector {
+    samples: BinaryHeap<Reverse<(i64, Vec<Vec<u8>>)>>,
+    null_count: Vec<i64>,
+    count: u64,
+    max_sample_size: usize,
+    fm_sketches: Vec<FmSketch>,
+    rng: StdRng,
+    total_sizes: Vec<i64>,
+    row_buf: Vec<u8>,
+}
+
+impl Default for RowSampleCollector {
+    fn default() -> Self {
+        RowSampleCollector {
+            samples: BinaryHeap::new(),
+            null_count: vec![],
+            count: 0,
+            max_sample_size: 0,
+            fm_sketches: vec![],
+            rng: StdRng::from_entropy(),
+            total_sizes: vec![],
+            row_buf: Vec::new(),
+        }
+    }
+}
+
+impl RowSampleCollector {
+    fn new(
+        max_sample_size: usize,
+        max_fm_sketch_size: usize,
+        col_and_group_len: usize,
+    ) -> RowSampleCollector {
+        RowSampleCollector {
+            samples: BinaryHeap::new(),
+            null_count: vec![0; col_and_group_len],
+            count: 0,
+            max_sample_size,
+            fm_sketches: vec![FmSketch::new(max_fm_sketch_size); col_and_group_len],
+            rng: StdRng::from_entropy(),
+            total_sizes: vec![0; col_and_group_len],
+            row_buf: Vec::new(),
+        }
+    }
+
+    pub fn collect_column_group(
+        &mut self,
+        columns_val: &Vec<Vec<u8>>,
+        column_groups: &Vec<tipb::AnalyzeColumnGroup>,
+    ) {
+        self.row_buf.clear();
+        let col_len = columns_val.len();
+        for i in 0..column_groups.len() {
+            let offsets = column_groups[i].get_column_offsets();
+            let mut all_null = true;
+            for j in 0..offsets.len() {
+                if columns_val[j][0] != NIL_FLAG {
+                    all_null = false
+                }
+                self.total_sizes[col_len + i] += columns_val[i].len() as i64
+            }
+            if all_null {
+                self.null_count[col_len + i] += 1;
+            }
+            // Use a in place murmur3 to replace this memory copy.
+            for j in 0..offsets.len() {
+                self.row_buf.extend_from_slice(&columns_val[j]);
+            }
+            self.fm_sketches[col_len + i].insert(&self.row_buf);
+        }
+    }
+
+    pub fn collect_column(&mut self, columns_val: Vec<Vec<u8>>) {
+        for i in 0..columns_val.len() {
+            if columns_val[i][0] == NIL_FLAG {
+                self.null_count[i] += 1;
+                return;
+            }
+            self.fm_sketches[i].insert(&columns_val[i]);
+            self.total_sizes[i] += columns_val[i].len() as i64;
+        }
+        self.sampling(columns_val);
+    }
+
+    pub fn sampling(&mut self, data: Vec<Vec<u8>>) {
+        let cur_rng = self.rng.gen_range(0, i64::MAX);
```

Comment:
Could you explain how this correlates to https://en.wikipedia.org/wiki/Reservoir_sampling#Algorithm_A-Res ? 
```
r := random() ^ (1/S.Weight)
```
Are we assigning the same weight to all rows?

Reply:
I think this is just the range interval is different. One is (0, 1), the other is [0, MaxInt64].
The range interval should not affect too much.

> Are we assigning the same weight to all rows?

We are assigning random value to each row.

Reply:
What confused me is `^ (1/S.Weight)` part. From the code, I guess `1/S.Weight` is selected to be 1 in this case?

Reply:
Yes, we choose this one since assigning random values can guarantee randomness in distributed cases.